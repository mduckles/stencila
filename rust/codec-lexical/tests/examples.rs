use std::{fs::read_to_string, path::PathBuf};

use codec::{
    common::{eyre::Result, glob::glob},
    format::Format,
    EncodeOptions,
};

use codec_lexical::{decode, encode};
use common_dev::insta::{assert_json_snapshot, assert_snapshot, assert_yaml_snapshot};

/// Decode each example of a Lexical document and create JSON and Lexical snapshots
/// including snapshots for losses
#[test]
fn examples() -> Result<()> {
    let pattern = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/examples")
        .canonicalize()?
        .to_string_lossy()
        .to_string()
        + "/**/*";

    for path in glob(&format!("{pattern}.lexical"))?
        .chain(glob(&format!("{pattern}.koenig"))?)
        .flatten()
    {
        let format = Format::from_path(&path);
        let ext = format.extension();

        let name = path
            .file_name()
            .map(|name| name.to_string_lossy())
            .and_then(|name| name.strip_suffix(&[".", &ext].concat()).map(String::from))
            .expect("should have format suffix");

        let original = read_to_string(path)?;

        let (article, info) = decode(&original, None)?;

        assert_json_snapshot!(format!("{name}.json"), article);
        assert_yaml_snapshot!(format!("{name}.decode.losses"), info.losses);

        let (lexical, info) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format.clone()),
                compact: Some(false),
                ..Default::default()
            }),
        )?;

        assert_snapshot!(format!("{name}.{ext}"), lexical);
        assert_yaml_snapshot!(format!("{name}.encode.losses"), info.losses);

        let (lexical, ..) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format),
                standalone: Some(true),
                compact: Some(false),
                ..Default::default()
            }),
        )?;
        assert_snapshot!(format!("{name}.standalone.{ext}"), lexical);
    }

    Ok(())
}
