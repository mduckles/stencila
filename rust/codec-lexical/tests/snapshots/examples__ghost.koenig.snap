---
source: rust/codec-lexical/tests/examples.rs
expression: redact(&lexical)
---
{
  "root": {
    "type": "root",
    "children": [
      {
        "type": "paragraph",
        "children": [
          {
            "type": "extended-text",
            "format": 0,
            "text": "A document for testing Stencila's integration with Ghost. This document is used as a test fixture, so be careful making changes because it is likely to affect those tests. "
          }
        ]
      },
      {
        "type": "paragraph",
        "children": [
          {
            "type": "extended-text",
            "format": 0,
            "text": "Paragraph with "
          },
          {
            "type": "extended-text",
            "format": 1,
            "text": "bold"
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": " "
          },
          {
            "type": "extended-text",
            "format": 2,
            "text": "italic"
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": " "
          },
          {
            "type": "extended-text",
            "format": 3,
            "text": "bold and italic"
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": " "
          },
          {
            "type": "extended-text",
            "format": 4,
            "text": "strikethrough"
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": " "
          },
          {
            "type": "extended-text",
            "format": 8,
            "text": "underline"
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": " "
          },
          {
            "type": "extended-text",
            "format": 16,
            "text": "code"
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": " "
          },
          {
            "type": "extended-text",
            "format": 32,
            "text": "subscript"
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": " "
          },
          {
            "type": "extended-text",
            "format": 64,
            "text": "superscript"
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": " "
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": "highlight"
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": " and a "
          },
          {
            "type": "extended-text",
            "format": 0,
            "text": "LOST Link"
          }
        ]
      },
      {
        "type": "extended-heading",
        "tag": "h1",
        "children": [
          {
            "type": "extended-text",
            "format": 0,
            "text": "Heading 1"
          }
        ]
      },
      {
        "type": "extended-heading",
        "tag": "h2",
        "children": [
          {
            "type": "extended-text",
            "format": 0,
            "text": "Heading 2"
          }
        ]
      },
      {
        "type": "extended-heading",
        "tag": "h3",
        "children": [
          {
            "type": "extended-text",
            "format": 0,
            "text": "Heading 3"
          }
        ]
      },
      {
        "type": "paragraph",
        "children": []
      },
      {
        "type": "extended-quote",
        "children": [
          {
            "type": "extended-text",
            "format": 0,
            "text": "A quote"
          }
        ]
      },
      {
        "type": "paragraph",
        "children": [
          {
            "type": "extended-text",
            "format": 0,
            "text": "LOST Unknown (aside)"
          }
        ]
      },
      {
        "type": "paragraph",
        "children": [
          {
            "type": "extended-text",
            "format": 0,
            "text": "LOST List"
          }
        ]
      },
      {
        "type": "paragraph",
        "children": [
          {
            "type": "extended-text",
            "format": 0,
            "text": "LOST List"
          }
        ]
      },
      {
        "type": "horizontalrule"
      },
      {
        "type": "image",
        "src": "https://stencila.ghost.io/content/images/2025/01/stencila-128.png"
      },
      {
        "type": "codeblock",
        "code": "print(\"Hello world\")",
        "language": "Python"
      },
      {
        "type": "markdown",
        "markdown": "| A   | B   |\n| --- | --- |\n| 1   | 2   |\n"
      },
      {
        "type": "html",
        "html": "<span>HTML card</span>"
      }
    ]
  }
}
